File: backend/app/network/protocol.py
import json
import zlib
import time
import logging
from typing import Dict, List, Optional, Tuple, Any

logger = logging.getLogger("ResQMesh.Protocol")

PACKET_TYPE_DATA = "DATA"
PACKET_TYPE_ACK = "ACK"
PACKET_TYPE_HEARTBEAT = "HEARTBEAT"


def compute_checksum(payload_bytes: bytes) -> str:
    """Compute CRC32 checksum hex string for data integrity verification."""
    return f"{zlib.crc32(payload_bytes) & 0xffffffff:08x}"


class Packet:
    """Transport-agnostic packet envelope with sequence numbering and checksums."""

    def __init__(
        self,
        packet_id: str,
        sender_id: str,
        packet_type: str = PACKET_TYPE_DATA,
        seq_num: int = 1,
        total_frags: int = 1,
        frag_idx: int = 0,
        payload: Optional[dict] = None,
        checksum: Optional[str] = None,
        timestamp: Optional[float] = None,
    ):
        self.packet_id = packet_id
        self.sender_id = sender_id
        self.packet_type = packet_type
        self.seq_num = seq_num
        self.total_frags = total_frags
        self.frag_idx = frag_idx
        self.payload = payload or {}
        self.timestamp = timestamp or time.time()

        if checksum:
            self.checksum = checksum
        else:
            self.checksum = self._generate_checksum()

    def _generate_checksum(self) -> str:
        data_str = json.dumps(self.payload, sort_keys=True)
        return compute_checksum(data_str.encode("utf-8"))

    def verify_checksum(self) -> bool:
        return self.checksum == self._generate_checksum()

    def to_dict(self) -> dict:
        return {
            "packet_id": self.packet_id,
            "sender_id": self.sender_id,
            "packet_type": self.packet_type,
            "seq_num": self.seq_num,
            "total_frags": self.total_frags,
            "frag_idx": self.frag_idx,
            "checksum": self.checksum,
            "timestamp": self.timestamp,
            "payload": self.payload,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Packet":
        pkt = cls(
            packet_id=data["packet_id"],
            sender_id=data["sender_id"],
            packet_type=data.get("packet_type", PACKET_TYPE_DATA),
            seq_num=data.get("seq_num", 1),
            total_frags=data.get("total_frags", 1),
            frag_idx=data.get("frag_idx", 0),
            payload=data.get("payload", {}),
            checksum=data.get("checksum"),
            timestamp=data.get("timestamp"),
        )
        return pkt


class PacketFramingParser:
    """Encodes packets into length-prefixed bytes streams and decodes stream chunks."""

    @staticmethod
    def encode_frame(packet: Packet) -> bytes:
        payload_bytes = json.dumps(packet.to_dict()).encode("utf-8")
        length = len(payload_bytes)
        # 4-byte big-endian header for frame length + payload
        return length.to_bytes(4, byteorder="big") + payload_bytes

    def __init__(self):
        self._buffer = bytearray()

    def feed_bytes(self, chunk: bytes) -> List[Packet]:
        """Feed bytes chunk and extract complete framed packets."""
        self._buffer.extend(chunk)
        packets = []

        while len(self._buffer) >= 4:
            frame_len = int.from_bytes(self._buffer[:4], byteorder="big")
            if len(self._buffer) < 4 + frame_len:
                break  # Wait for full payload

            payload_bytes = bytes(self._buffer[4 : 4 + frame_len])
            del self._buffer[: 4 + frame_len]

            try:
                data = json.loads(payload_bytes.decode("utf-8"))
                pkt = Packet.from_dict(data)
                if pkt.verify_checksum():
                    packets.append(pkt)
                else:
                    logger.warning(f"Checksum mismatch for packet {pkt.packet_id}, dropping.")
            except Exception as e:
                logger.error(f"Error parsing packet payload frame: {e}")

        return packets


class ACKTracker:
    """Manages pending unacknowledged packets and retransmission tracking."""

    def __init__(self, sender_id: str, timeout_seconds: float = 5.0, max_retries: int = 3):
        self.sender_id = sender_id
        self.timeout_seconds = timeout_seconds
        self.max_retries = max_retries
        self.seq_counter = 0
        self.pending_acks: Dict[str, dict] = {}

    def get_next_seq_num(self) -> int:
        self.seq_counter += 1
        return self.seq_counter

    def track_packet(self, packet: Packet) -> None:
        if packet.packet_type == PACKET_TYPE_DATA:
            self.pending_acks[packet.packet_id] = {
                "packet": packet,
                "sent_at": time.time(),
                "retries": 0,
            }

    def process_ack(self, ack_packet: Packet) -> Optional[str]:
        if ack_packet.packet_type == PACKET_TYPE_ACK:
            acked_id = ack_packet.payload.get("acked_packet_id")
            if acked_id in self.pending_acks:
                del self.pending_acks[acked_id]
                logger.info(f"Received ACK for packet: {acked_id}")
                return acked_id
        return None

    def get_packets_due_for_retransmission(self) -> List[Packet]:
        now = time.time()
        due: List[Packet] = []
        expired_ids: List[str] = []

        for pkt_id, item in self.pending_acks.items():
            if now - item["sent_at"] >= self.timeout_seconds:
                if item["retries"] < self.max_retries:
                    item["retries"] += 1
                    item["sent_at"] = now
                    due.append(item["packet"])
                    logger.info(f"Retransmitting packet {pkt_id} (Attempt {item['retries']})")
                else:
                    expired_ids.append(pkt_id)
                    logger.warning(f"Packet {pkt_id} reached max retries, dropping.")

        for pkt_id in expired_ids:
            del self.pending_acks[pkt_id]

        return due

    def create_ack_packet(self, data_packet: Packet) -> Packet:
        return Packet(
            packet_id=f"ack-{data_packet.packet_id}",
            sender_id=self.sender_id,
            packet_type=PACKET_TYPE_ACK,
            payload={"acked_packet_id": data_packet.packet_id, "acked_seq_num": data_packet.seq_num},
        )

File: mobile/src/network/protocol.ts
/**
 * ResQMesh Reliable Messaging Protocol for React Native Mobile Client.
 * Implements CRC32 checksums, sequence numbering, stream framing, and ACK tracking.
 */

export const PACKET_TYPE_DATA = 'DATA';
export const PACKET_TYPE_ACK = 'ACK';
export const PACKET_TYPE_HEARTBEAT = 'HEARTBEAT';

export interface PacketData {
  packet_id: string;
  sender_id: string;
  packet_type: string;
  seq_num: number;
  total_frags: number;
  frag_idx: number;
  checksum: string;
  timestamp: number;
  payload: Record<string, any>;
}

// Basic CRC32 implementation for React Native / TS
export function computeCRC32(str: string): string {
  let crc = 0 ^ (-1);
  for (let i = 0; i < str.length; i++) {
    const code = str.charCodeAt(i);
    crc = (crc >>> 8) ^ crcTable[(crc ^ code) & 0xff];
  }
  const result = (crc ^ (-1)) >>> 0;
  return result.toString(16).padStart(8, '0');
}

const crcTable: number[] = (() => {
  const table: number[] = [];
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) {
      c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    }
    table[n] = c;
  }
  return table;
})();

export class Packet {
  public packetId: string;
  public senderId: string;
  public packetType: string;
  public seqNum: number;
  public totalFrags: number;
  public fragIdx: number;
  public checksum: string;
  public timestamp: number;
  public payload: Record<string, any>;

  constructor(opts: {
    packetId: string;
    senderId: string;
    packetType?: string;
    seqNum?: number;
    totalFrags?: number;
    fragIdx?: number;
    payload?: Record<string, any>;
    checksum?: string;
    timestamp?: number;
  }) {
    this.packetId = opts.packetId;
    this.senderId = opts.senderId;
    this.packetType = opts.packetType || PACKET_TYPE_DATA;
    this.seqNum = opts.seqNum || 1;
    this.totalFrags = opts.totalFrags || 1;
    this.fragIdx = opts.fragIdx || 0;
    this.payload = opts.payload || {};
    this.timestamp = opts.timestamp || Date.now() / 1000;

    if (opts.checksum) {
      this.checksum = opts.checksum;
    } else {
      this.checksum = this.generateChecksum();
    }
  }

  public generateChecksum(): string {
    const sortedStr = JSON.stringify(this.payload, Object.keys(this.payload).sort());
    return computeCRC32(sortedStr);
  }

  public verifyChecksum(): boolean {
    return this.checksum === this.generateChecksum();
  }

  public toDict(): PacketData {
    return {
      packet_id: this.packetId,
      sender_id: this.senderId,
      packet_type: this.packetType,
      seq_num: this.seqNum,
      total_frags: this.totalFrags,
      frag_idx: this.fragIdx,
      checksum: this.checksum,
      timestamp: this.timestamp,
      payload: this.payload,
    };
  }

  public static fromDict(data: PacketData): Packet {
    return new Packet({
      packetId: data.packet_id,
      senderId: data.sender_id,
      packetType: data.packet_type,
      seqNum: data.seq_num,
      totalFrags: data.total_frags,
      fragIdx: data.frag_idx,
      checksum: data.checksum,
      timestamp: data.timestamp,
      payload: data.payload,
    });
  }
}

export class ACKTrackerMobile {
  private senderId: string;
  private seqCounter: number = 0;
  private pendingAcks: Map<string, { packet: Packet; sentAt: number; retries: number }> = new Map();
  private timeoutMs: number;
  private maxRetries: number;

  constructor(senderId: string, timeoutMs: number = 5000, maxRetries: number = 3) {
    this.senderId = senderId;
    this.timeoutMs = timeoutMs;
    this.maxRetries = maxRetries;
  }

  public getNextSeqNum(): number {
    this.seqCounter += 1;
    return this.seqCounter;
  }

  public trackPacket(packet: Packet): void {
    if (packet.packetType === PACKET_TYPE_DATA) {
      this.pendingAcks.set(packet.packetId, {
        packet,
        sentAt: Date.now(),
        retries: 0,
      });
    }
  }

  public processAck(ackPacket: Packet): string | null {
    if (ackPacket.packetType === PACKET_TYPE_ACK) {
      const ackedId = ackPacket.payload['acked_packet_id'];
      if (ackedId && this.pendingAcks.has(ackedId)) {
        this.pendingAcks.delete(ackedId);
        return ackedId;
      }
    }
    return null;
  }

  public getDueRetransmissions(): Packet[] {
    const now = Date.now();
    const due: Packet[] = [];

    this.pendingAcks.forEach((item, pktId) => {
      if (now - item.sentAt >= this.timeoutMs) {
        if (item.retries < this.maxRetries) {
          item.retries += 1;
          item.sentAt = now;
          due.push(item.packet);
        } else {
          this.pendingAcks.delete(pktId);
        }
      }
    });

    return due;
  }

  public createAckPacket(dataPacket: Packet): Packet {
    return new Packet({
      packetId: `ack-${dataPacket.packetId}`,
      senderId: this.senderId,
      packetType: PACKET_TYPE_ACK,
      payload: {
        acked_packet_id: dataPacket.packetId,
        acked_seq_num: dataPacket.seqNum,
      },
    });
  }
}
